src/gui/shell.cpp

```diff
@@ -161,12 +143,96 @@ bool Shell::setGuiFont(const QString& fdesc, bool force, bool updateOption, bool
 	// Font changed: trigger resize to update the ShellWidget, and update font variables.
 	resizeNeovim(size());
 
-	m_nvim->api0()->vim_set_var("GuiFont", fontDesc());
+	m_nvim->api0()->vim_set_var("GuiFont", fontDesc(false));
 
 	// Updating guifont when the user has already called 'set guifont=...' may cause
 	// unwanted recursion. Only update this option for ':GuiFont', and dialog calls.
 	if (isGuiDialogRequest || updateOption) {
-		m_nvim->api0()->vim_set_option("guifont", fontDesc());
+		m_nvim->api0()->vim_set_option("guifont", fontDesc(false));
+	}
+
+	return true;
+}
+
+bool Shell::setGuiFontWide(const QString& fdesc, bool force, bool updateOption)
```

Comment:
A lot of this logic is duplicated from `setGuiFont`, but no longer necessary. 

We should get rid of `GuiFontWide`, so the update logic is not necessary.
We should consider getting rid of `force`.
I'm not sure how the font dialog works with a list of fonts, it should also be removed.

This feature should mimic gvim.

Reply:
Also... Fine for now to get things working.

However, we shouldn't copy-paste code for the final iteration. Instead we should move the re-used parts to functions. Just a friendly reminder :)

Reply:
> Yes, I think there needs to be a std::vector<QFont> ShellWidget::m_fontWide;. Or maybe m_fontWideList?

I use `QVector<QFont>` to store the wide font list. Is it neccessary use `std::vector`?
With using `m_fontWideList`, I add `ShellWidget::createShellFont` and I think the `ShellWidget::setShellFontWide(const QVector<QFont>& font)` is required. 

I'm agree that re-using code, and there `Shell::getFontFromDesc` was add to handle all font description with help of `ShellWidget::GetQFrontFromDescription(...)` .

Reply:
> We then need some additional logic in the rendering event (pseudocode):

`QFontMetrics` is used to check whether wide font contains double-character. But it's seemed that `QFontMetrics(non-wide font).inFontUcs4` return `true` too. Is it possible that `QFontMetrics` find a default font from system for double-character? 

```
if (cell.IsDoubleWidth()) {
	for (const auto& fw : m_fontWideList) {
		if (QFontMetrics(fw).inFontUcs4(cell.GetCharacter())) {
			f = QFont(fw);
			break;
		}
	}
} else {
        ...
}
```

Reply:
> I use QVector<QFont> to store the wide font list. Is it neccessary use std::vector?

No, Qt containers work fine too. I think the optimal Qt container for this case is `QList`, we don't need the memory continuity guarantee of `QVector`.

> But it's seemed that QFontMetrics(non-wide font).inFontUcs4 return true too.

Interesting... I'm not familiar enough with the inner workings of QFontMetrics. We'll have to investigate this.

Reply:
`QList` was used to store wide fonts. :)